### PR TITLE
Authentication updates (frontend + google auth + cleanups) + graphql field validations

### PR DESCRIPTION
**Pull request by pouriaMaleki**
To test this:

Make your own version of .env in the backend folder. (use .env.sample file)
Start frontend and backend using `npm start` in each folder and go to `localhost:4000/login`.

When testing the "forgot password", because we don't have an email service yet, you can see the token logged in the backend console.

To test google login, you will need to contact me and tell me your Gmail address. This is because this client is not yet approved by google. (Alternatively, you can create an oauth2 key in your own google console, and provide the client key in .env files of frontend and backend)

### CODE REVIEW

**Comment by albacoretuna** [comment]
Thanks for all the work here :)

**Comment by albacoretuna** [comment]
Is this still needed? I needed to npm install it before backend worked. 

Error: Cannot find module 'astral-regex'

**Review by albacoretuna — COMMENTED** [review]
Wow this is epic :) 
Looks great. I have a couple of small concerns about the password reset. 

1. The tokens I got was like 2341234. Shouldn't it be a bit stronger? like UUID v4 etc.? 
2. When we send the forgot password email eventually, the users will get a link like this:
getmoro.com/reset-password?email=olo@example.com&token=234234

This is slightly problematic both from GDPR and security perspective. Because personally identified informations shouldn't be in links, as they will be reported to web analytics and will be logged there. Secondly, now we are giving a chance for brute force. Just ping that domain name with any email and token and something might work! 

Ideally our password reset link should be an id, and a token. The id can be the permanent user id, or something temporary.

**Comment by pouriaMaleki** [comment]
> 1. The tokens I got was like 2341234. Shouldn't it be a bit stronger? like UUID v4 etc.?
> 2. When we send the forgot password email eventually, the users will get a link like this:
>    getmoro.com/reset-password?email=[olo@example.com](mailto:olo@example.com)&token=234234
>
> This is slightly problematic both from GDPR and security perspective. Because personally identified informations shouldn't be in links, as they will be reported to web analytics and will be logged there. Secondly, now we are giving a chance for brute force. Just ping that domain name with any email and token and something might work!

@albacoretuna Thanks for reviewing. I haven't planed to make and send a reset password link in the email, instead, the user has to enter the 6 digit token received in the email manually on the reset password page.
Good catch on the brute force, I'll make some changes that deny the password reset after entering the wrong token for the third time. Plus limiting the user from asking for a password reset token too many times.

**Comment by albacoretuna** [comment]
> 
> @albacoretuna Thanks for reviewing. I haven't planed to make and send a reset password link in the email, instead, the user has to enter the 6 digit token received in the email manually on the reset password page.
> Good catch on the brute force, I'll make some changes that deny the password reset after entering the wrong token for the third time. Plus limiting the user from asking for a password reset token too many times.

This works too, just be warned that in once company we had to change it from PIN to link. Someone on stackoverflow explains it better than I can: 

>The reason I encourage you to reconsider this is because your alternative (using a PIN) is going to be more complicated for users and will likely cause more headaches for you. The PIN complexity aside for a moment (numbers only => insecure, letters and numbers => more complicated), you are going to have to instruct users on what to do, and these instructions will necessitate you telling them where on your site to go to enter the code. While saying "Go to our site home page, click Forgotten Password, and click Enter Reset PIN" isn't rocket science you will still have users complain that they can't find the buttons or otherwise get confused about what to do. Having them simply click on a link eliminates a lot of this confusion.

**Comment by pouriaMaleki** [comment]
@albacoretuna I have pushed the changes to fix the security concerns with having only numbers, but after that, I saw your recent comment. Fair point. Thanks.

 I'll make the changes to create one big unique identifier (UUID v4) and will put it in a link to the `/passwordReset` page. (It probably will be tomorrow)

**Comment by albacoretuna** [comment]
> @albacoretuna I have pushed the changes to fix the security concerns with having only numbers, but after that, I saw your recent comment. Fair point. Thanks.
> 
> I'll make the changes to create one big unique identifier (UUID v4) and will put it in a link to the `/passwordReset` page. (It probably will be tomorrow)


@pouriaMaleki  You can also merge it as is. We don't have to do it now, we can always come back to it later :) This is already big step forward as is.

**Comment by pouriaMaleki** [comment]
> @pouriaMaleki You can also merge it as is. We don't have to do it now, we can always come back to it later :) This is already a big step forward as is.

Thanks @albacoretuna, this already has so many loose ends, I'd better keep it clean. Changes are applied now.

File: backend/src/user/login.ts [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+      data: {
+        email: credentials.email,
+        resolved: false,
+      },
+    });
+
+    return { success: false, message: 'Incorrect credentials' };
+  }
+
+  // resolve failed attempts
```
Comment (pouriaMaleki) [inline backend/src/user/login.ts:51]: Maybe include an identifier like IP or something to be sure it's the same person who logged in.

**Pull request merged**